lib/events/session_writer.go

```diff
@@ -461,38 +462,31 @@ func (a *SessionWriter) processEvents() {
 	}
 }
 
-// IsPermanentEmitError checks if the error contains underlying BadParameter error.
+// IsPermanentEmitError checks if the error contains either a sole
+// [trace.BadParameter] error in its chain, or a [trace.Aggregate] error
+// composed entirely of BadParameters.
 func IsPermanentEmitError(err error) bool {
-	var (
-		maxDeep            = 50
-		iter               = 0
-		isPerErrRecurCheck func(error) bool
-	)
-
-	isPerErrRecurCheck = func(err error) bool {
-		defer func() { iter++ }()
-		if iter >= maxDeep {
-			return false
-		}
+	return isPermanentEmitError(err, 1 /* depth */)
+}
 
-		if trace.IsBadParameter(err) {
-			return true
-		}
-		if !trace.IsAggregate(err) {
-			return false
-		}
-		agg, ok := trace.Unwrap(err).(trace.Aggregate)
-		if !ok {
-			return false
-		}
-		for _, err := range agg.Errors() {
-			if !isPerErrRecurCheck(err) {
+func isPermanentEmitError(err error, depth int) bool {
```

Comment:
@smallinsky, this is the equivalent implementation for IsPermanentEmitError with the new trace. It clears the test, but it looks kind of strange to me.

Is there a particular reason why aggregates, specifically, must be entirely composed of BadParameters? If not, then could I replace this func with a `return trace.IsBadParameter(err)` ?